SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestReplicatedStartupActor/SpatialTestReplicatedStartupActor.cpp

```diff
@@ -44,36 +54,143 @@ void ASpatialTestReplicatedStartupActor::BeginPlay()
 
-			AReplicatedStartupActorPlayerController* PlayerController = Cast<AReplicatedStartupActorPlayerController>(GetLocalFlowController()->GetOwner());
+	// The server sets default values for the replicated properties.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorServerSetDefaultProperties"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest)
+		{
+			ReplicatedStartupActor->TestIntProperty = 1;
+
+			ReplicatedStartupActor->TestArrayProperty.Empty();
+			ReplicatedStartupActor->TestArrayProperty.Add(1);
 
-			PlayerController->ClientToServerRPC(this, ReplicatedActors[0]);
+			ReplicatedStartupActor->TestArrayStructProperty.Empty();
+			ReplicatedStartupActor->TestArrayStructProperty.Add(FTestStruct{1});
 
 			FinishStep();
 		});
 
-	AddStep(TEXT("SpatialTestReplicatedStarupActorClientsCheckStep"), FWorkerDefinition::AllClients, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+	// All workers check that the properties were replicated correctly.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorAllWorkersCheckDefaultProperties"), FWorkerDefinition::AllWorkers, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
 		{
-			if (bIsValidReference)
+			if (ReplicatedStartupActor->TestIntProperty == 1
+				&& ReplicatedStartupActor->TestArrayProperty.Num() == 1 && ReplicatedStartupActor->TestArrayProperty[0] == 1
+				&& ReplicatedStartupActor->TestArrayStructProperty.Num() == 1 && ReplicatedStartupActor->TestArrayStructProperty[0].Int == 1)
 			{
-				AssertTrue(bIsValidReference, TEXT("The server has a valid reference to this client's replicated actor"));
+				FinishStep();
+			}
+		}, 5.0f);
 
-				AReplicatedStartupActorPlayerController* PlayerController = Cast<AReplicatedStartupActorPlayerController>(GetLocalFlowController()->GetOwner());
-				PlayerController->ResetBoolean(this);
+	// The server moves the ReplicatedStartupActor out of the clients' view.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorServerMoveActorOutOfView"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest)
+		{
+			ReplicatedStartupActor->SetActorLocation(FVector(15000.0f, 15000.0f, 50.0f));
+
+			FinishStep();
+		});
 
+	// All workers check that the movement is visible.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorAllWorkersCheckMovement"), FWorkerDefinition::AllWorkers, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+		{
+			// Make sure the Actor was moved out of view of the clients before updating its properties
+			if (ReplicatedStartupActor->GetActorLocation().Equals(FVector(15000.0f, 15000.0f, 50.0f), 1))
+			{
 				FinishStep();
 			}
-		}, 2.0);
+		}, 5.0f);
+
+	// The server updates the replicated properties whilst the ReplicatedStartupActor is out of the clients' view.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorServerUpdateProperties"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest)
```

Comment:
What is the purpose of moving the actor out of view?

Reply:
I'm not sure if this is currently relevant, however this was taken from one of Miron's old tests, so maybe there was a previous bug that prevented this from happening?

Reply:
Yeah, this is for UNR-3128 - emptying an array while the bNetLoadOnClient actor is out-of-view triggers the bug (after moving back into view).

Reply:
Oh, I completely forgot about actually documenting this in the test itself 😄 . It's actually a couple of lines under the step definition.